File: 532-k-diff-pairs-in-an-array/532-k-diff-pairs-in-an-array.py
class Solution:
    def findPairs(self, nums: List[int], k: int) -> int:
        sol = set()
        numsSet = set()
        
        for i in range(len(nums)):
            if nums[i] in numsSet and k != 0:
                continue
            a, b, c = nums[i], nums[i] + k, nums[i] - k
            if b in numsSet and (a, b) not in sol and (b, a) not in sol:
                sol.add((a, b))
            if c in numsSet and (a, c) not in sol and (c, a) not in sol:
                sol.add((a, c))
            numsSet.add(a)
        return len(sol)

File: 532-k-diff-pairs-in-an-array/532-k-diff-pairs-in-an-array.c
#define MAX(a, b) (((a) > (b)) ? (a) : (b))

int Comparator(const void *a , const void *b){
    return *(int *)a - *(int *)b;
}


int findPairs(int* nums, int numsSize, int k){
    int i = 0, j = 0, res = 0;
    qsort(nums, numsSize, sizeof(int), Comparator);
    
    while(i < numsSize){
        
        while( i > 0 && i < numsSize && nums[i] == nums[i - 1])
            i++;
        
        j = MAX(j, i + 1);
        while(j < numsSize && nums[j] < nums[i] + k)
            j++;
        
        if(j >= numsSize)
            return res;
        
        if(nums[j] == nums[i] + k){
            res++;
            j++;
        }
        i++;
    }
    
    return res;
}